### ISSUE TITLE
Prestashop incorrectly collects data about user activity (window unload event is unreliable)

### ISSUE THREAD

**Issue by ProteanCode** [PrestaShop/PrestaShop#20921]
#### Describe the bug

Prestashop module named statsdata uses an unload event to send a POST request with data about user activity. This does not work for me since I get a 99% of 499s because web browser cannot handle the request fully.

A shop is hosted on dedicated machine with good hardware and network, it depends more on a client network speed, but the unload event may not be reliable for this purpose simply because we have multiple web browsers.

#### Expected behavior

I expect to receive a HTTP 200 response code, not a 499

#### Steps to Reproduce
>Edit by Khouloud
1. Go to BO > Module manager page > search for the `statsdata` module and configure it.
2. Enable `Save page views for each customer`,  `Save global page views options` and save
3. Go to FO > Inspect the element > in the Network tab > Select the "Preserve log" option and reload the page
4. see an error
![image](https://user-images.githubusercontent.com/16067358/93072766-267b0800-f682-11ea-8b90-0aaa22a56bc4.png)

1. Go to any shop page
2. Close tab / change url
3. At this stage the unload event should be triggered
4. The web browser (brave in my case which is chrome based) have no time to handle the post request to the server, therefore server logs 499 http code

#### Additional information

```
traefik2@myserver    | <customer_ip> - - [08/Sep/2020:22:55:14 +0000] "POST /index.php?controller=statistics HTTP/2.0" 499 21 "-" "-" 5350380 "project-secured@docker" "http://<proxy_ip>:80" 14ms
traefik2@myserver    | <customer_ip> - - [08/Sep/2020:22:57:48 +0000] "POST /index.php?controller=statistics HTTP/2.0" 499 21 "-" "-" 5350974 "project-secured@docker" "http://<proxy_ip>:80" 0ms
traefik2@myserver    | <customer_ip> - - [08/Sep/2020:22:57:50 +0000] "POST /index.php?controller=statistics HTTP/2.0" 499 21 "-" "-" 5350976 "project-secured@docker" "http://<proxy_ip>:80" 0ms
```

* repository: https://github.com/PrestaShop/statsdata/blob/4de78d4c0b096f2d6768f5c96856ea8eabd8358c/statsdata.php#L127
* PrestaShop version: 1.7.7.0
* PHP version: 7.2
* Stackoverflow similar issue: https://stackoverflow.com/questions/21228349/sending-a-post-request-with-javascript-on-unload-beforeunload-is-that-possible

**Comment by khouloudbelguith** [comment]
Hi @ProteanCode,

Thanks for your report.
>Close tab / change url

At this step, you are requesting a page not found, isn't? this event is not saved in your case? is it?
In my case, with PS1.7.6.7, I don't have any 499 errors in the var/logs/log file.

Thanks!

**Comment by ProteanCode** [comment]
> Hi @ProteanCode,
> 
> Thanks for your report.
> 
> > Close tab / change url
> 
> At this step, you are requesting a page not found, isn't? this event is not saved in your case? is it?
> In my case, with PS1.7.6.7, I don't have any 499 errors in the var/logs/log file.
> 
> Thanks!

I could be a bit unprecise about this. Its better to read the:
https://developer.mozilla.org/en-US/docs/Web/API/Window/unload_event
and
https://api.jquery.com/jquery.post/

Afaik using this method is okayish when called code is synchronous, but the` jquery.post()` is not

There are few more things you could do to reproduce it:
* in chrome based web browser (maybe FF now supports it) change your network speed to something like "slow 3G"
* in web browser development tools check the "Preserve logs" so you can see the actualy request in Network tab
* navigate between links, the expected (incorrect) behaviour would be an aborted / cancelled request

This is not a 100% reproduction rate bug since I actually get 3 error codes: 200, 499, and 500 but I can believe the 500 is related to something else that does not indirectly targets this issue:

```
Uncaught  --> Smarty: 0():Missing '$template' parameter
thrown in /var/www/html/vendor/smarty/smarty/libs/sysplugins/smarty_internal_templatebase.php
```

Thanks for quick reply btw.

**Comment by ProteanCode** [comment]
I am also pointing out that this request should never be aborted in the first place

**Comment by ProteanCode** [comment]
I have also found an issues that may be related:
https://github.com/PrestaShop/PrestaShop/issues/12408#issuecomment-459785624
https://github.com/PrestaShop/statsdata/pull/7

**Comment by khouloudbelguith** [comment]
@ProteanCode,

Is it the same issue?
![image](https://user-images.githubusercontent.com/16067358/92618134-544a0080-f2c0-11ea-8098-ab7601f3a7bd.png)
as reported here previously
http://forge.prestashop.com/browse/BOOM-4826?

Thanks to check and feedback.

**Comment by ProteanCode** [comment]
> @ProteanCode,
> 
> Is it the same issue?
> ![image](https://user-images.githubusercontent.com/16067358/92618134-544a0080-f2c0-11ea-8098-ab7601f3a7bd.png)
> as reported here previously
> http://forge.prestashop.com/browse/BOOM-4826?
> 
> Thanks to check and feedback.

I have no access to this Jira ticket (the account is required), but yeah the status code is 500 so it looks like its the same case. Sometimes you won't get 500 because the request will be aborted / cancelled before its finish

**Comment by khouloudbelguith** [comment]
@ProteanCode, thanks for your feedback.
I manage to reproduce the issue with PS1.7.6.7 and PS1770beta2 using the `statsdata` module v2.1.0.
![image](https://user-images.githubusercontent.com/16067358/93072436-ab195680-f681-11ea-8644-65e9c62fe2b0.png)

I’ll add this to the debug roadmap.
Here's the PR: https://github.com/PrestaShop/statsdata/pull/7 to fix the issue.
Thanks!

**Comment by ProteanCode** [comment]
Thanks.

As I understand if I git clone the correct code from statsdata repository into /modules directory and clear the cache right?

No some additional install procedures?

**Comment by khouloudbelguith** [comment]
@ProteanCode, this issue is not fixed yet.
But you can check this patch: https://github.com/PrestaShop/statsdata/pull/7/files
But it is still not merged (not tested by our QA team yet).

Thanks!

**Comment by HartLarsson** [comment]
> @ProteanCode, this issue is not fixed yet.
> But you can check this patch: https://github.com/PrestaShop/statsdata/pull/7/files
> But it is still not merged (not tested by our QA team yet).
> 
> Thanks!

i've applied the patch as you suggested but continue to receive the same error:

`[13-Nov-2020 11:06:48 Europe/Rome] PHP Fatal error:  Uncaught  --> Smarty: 0():Missing '$template' parameter <-- 
  thrown in /home/www/public_html/vendor/smarty/smarty/libs/sysplugins/smarty_internal_templatebase.php on line 177
[13-Nov-2020 11:06:51 Europe/Rome] PHP Fatal error:  Uncaught  --> Smarty: 0():Missing '$template' parameter <-- 
  thrown in /home/www/public_html/vendor/smarty/smarty/libs/sysplugins/smarty_internal_templatebase.php on line 177`

PS 1.7.6.7
PHP7.2

thanks

**Comment by ProteanCode** [comment]
Is this going to be adressed in 1.7.7.x stable?
I saw some bigger updates like jQuery to v3 in FO

**Comment by raulrockin** [comment]
I have the same problem with 1.7.6.9
I also tryed the https://github.com/PrestaShop/statsdata/pull/7/files solution but the problem is still there.
The error is happening when customers are paying the order in the last step, and the error cancel the order because it returns an error 500

**Comment by yo791** [comment]
Hello,
I have the same problem with 1.7.6.9 php7.2 and i have applied the patch too. This error is happering on certain product and categories.
With the debug mode back office this error don't appear in logs.
This error appear with classic theme and custom theme, i have erase prod and dev in the /var/cache and this problem is not fixed.
The problem with this error is the shop is very low.

> [Thu Feb 25 10:05:38.131847 2021] [php7:error] [pid 6055] [client :26789] PHP Fatal error:  Uncaught  --> Smarty: 0():Missing '$template' parameter <-- \n  thrown in /home/www/vendor/smarty/smarty/libs/sysplugins/smarty_internal_templatebase.php on line 177

**Comment by Maofree** [comment]
Hi
from the log I get the same error (I use a dedicated server for this website)

`[Mon Sep 20 10:43:48.912496 2021] [proxy_fcgi:error] [pid 150105:tid 139779819620096] [client 108.162.245.72:25722] AH01071: Got error 'PHP message: PHP Fatal error:  Uncaught  --> Smarty: 0():Missing '$template' parameter <-- \n  thrown in /var/www/vhosts/kasa-store.com/public_html/arredamento-online/vendor/smarty/smarty/libs/sysplugins/smarty_internal_templatebase.php on line 177\n'`

bye

**Comment by ProteanCode** [comment]
This ticket is almost a year long.

Is there any concept on how to fix that? There are many coders around that could make their first attempt on making a change to this project, but they do not want to reverse engineer all the core concepts underneath.

Any alternative plugin that could work with statsdata turned off?

**Comment by PierreRambaud** [comment]
As far as I know, the problem comes from the fact that the request is started, and directly closed by the unload event.
The ajax request shouldn't be in async mode during the unload event. That's the reason why it's not working as expected.

Here is an example of how it should be handled:
https://stackoverflow.com/a/19275264

**Comment by metacreo** [comment]
Fixed in statsdata 2.1.1
